first step alphabet refactorisation

h-2 opened the pull request:
see #869 

This PR only does syntactical changes. I would like to get this merged before working on the rest so that later PRs only affect few files.

eseiler commented on include/seqan3/range/view/char_to.hpp:
```diff
@@ -52,7 +52,7 @@ namespace seqan3::view
  * | seqan3::const_iterable_concept  |                                       | *preserved*                                        |
  * |                                 |                                       |                                                    |
- * | seqan3::reference_t             | seqan3::underlying_char_t<alphabet_t> | `alphabet_t`                                       |
+ * | seqan3::reference_t             | seqan3::alphabet_char_t<alphabet_t> | `alphabet_t`                                       |
```
```suggestion
 * | seqan3::reference_t             | seqan3::alphabet_char_t<alphabet_t>   | `alphabet_t`                                       |
```

eseiler commented on include/seqan3/range/view/rank_to.hpp:
```diff
@@ -51,7 +51,7 @@ namespace seqan3::view
  * | seqan3::const_iterable_concept  |                                       | *preserved*                                        |
  * |                                 |                                       |                                                    |
- * | seqan3::reference_t             | seqan3::underlying_rank_t<alphabet_t> | `alphabet_t`                                       |
+ * | seqan3::reference_t             | seqan3::alphabet_rank_t<alphabet_t> | `alphabet_t`                                       |
```
```suggestion
* | seqan3::reference_t             | seqan3::alphabet_rank_t<alphabet_t>   | `alphabet_t`                                       |
```

eseiler commented on include/seqan3/range/view/to_char.hpp:
```diff
@@ -50,7 +50,7 @@ namespace seqan3::view
  * | seqan3::const_iterable_concept  |                                       | *preserved*                                        |
  * |                                 |                                       |                                                    |
- * | seqan3::reference_t             | seqan3::Alphabet                      | seqan3::underlying_char_t<seqan3::value_type_t<urng_t>> |
+ * | seqan3::reference_t             | seqan3::Alphabet                      | seqan3::alphabet_char_t<seqan3::value_type_t<urng_t>> |
```
```suggestion
 * | seqan3::reference_t             | seqan3::Alphabet                      | seqan3::alphabet_char_t<seqan3::value_type_t<urng_t>>   |
```

eseiler commented on include/seqan3/range/view/to_rank.hpp:
```diff
@@ -50,13 +50,13 @@ namespace seqan3::view
  * | seqan3::const_iterable_concept  |                                       | *preserved*                                        |
  * |                                 |                                       |                                                    |
- * | seqan3::reference_t             | seqan3::Alphabet                      | seqan3::underlying_rank_t<seqan3::value_type_t<urng_t>> |
+ * | seqan3::reference_t             | seqan3::Alphabet                      | seqan3::alphabet_rank_t<seqan3::value_type_t<urng_t>> |
```
```suggestion
 * | seqan3::reference_t             | seqan3::Alphabet                      | seqan3::alphabet_rank_t<seqan3::value_type_t<urng_t>>   |
```

eseiler commented on include/seqan3/alphabet/concept.hpp:
```diff
@@ -77,10 +77,10 @@ SEQAN3_CONCEPT Semialphabet = std::Regular<std::remove_reference_t<t>> &&
-    requires           noexcept(assign_rank(std::remove_reference_t<t>{}, 0));
-    requires std::Same<decltype(assign_rank(v,                            0)), std::remove_reference_t<t> &>;
-    requires std::Same<decltype(assign_rank(std::remove_reference_t<t>{}, 0)), std::remove_reference_t<t>  >;
+    requires           noexcept(assign_rank_to(                           0, v));
```
```suggestion
    requires           noexcept(assign_rank_to(0                           , v));
```

eseiler commented on include/seqan3/alphabet/concept.hpp:
```diff
@@ -77,10 +77,10 @@ SEQAN3_CONCEPT Semialphabet = std::Regular<std::remove_reference_t<t>> &&
-    requires std::Same<decltype(assign_rank(std::remove_reference_t<t>{}, 0)), std::remove_reference_t<t>  >;
+    requires           noexcept(assign_rank_to(                           0, v));
+    requires           noexcept(assign_rank_to(0, std::remove_reference_t<t>{}));
+    requires std::Same<decltype(assign_rank_to(                           0, v)), std::remove_reference_t<t> &>;
```
```suggestion
    requires std::Same<decltype(assign_rank_to(0                           , v)), std::remove_reference_t<t> &>;
```

eseiler commented on test/unit/alphabet/adaptation/char_test.cpp:
```diff
@@ -57,19 +57,19 @@ TYPED_TEST(char_adaptation, to_char)
-    EXPECT_TRUE((std::is_same_v<decltype(assign_char(TypeParam{'A'}, 'A')), alphabet_char_t<TypeParam>  >));
-    EXPECT_EQ((assign_char(TypeParam{'C'}, 'A')), l);
-    EXPECT_EQ((assign_char(l,              'C')), TypeParam{'C'});
+    EXPECT_TRUE((std::is_same_v<decltype(assign_char_to(             'A', l)), alphabet_char_t<TypeParam> &>));
```
```suggestion
    EXPECT_TRUE((std::is_same_v<decltype(assign_char_to('A', l             )), alphabet_char_t<TypeParam> &>));
```

eseiler commented on test/unit/alphabet/adaptation/char_test.cpp:
```diff
@@ -57,19 +57,19 @@ TYPED_TEST(char_adaptation, to_char)
+    EXPECT_TRUE((std::is_same_v<decltype(assign_char_to(             'A', l)), alphabet_char_t<TypeParam> &>));
+    EXPECT_TRUE((std::is_same_v<decltype(assign_char_to('A', TypeParam{'A'})), alphabet_char_t<TypeParam>  >));
+    EXPECT_EQ((assign_char_to('A', TypeParam{'C'})), l);
+    EXPECT_EQ((assign_char_to(             'C', l)), TypeParam{'C'});
```
```suggestion
    EXPECT_EQ((assign_char_to('C', l             )), TypeParam{'C'});
```

eseiler commented on test/unit/alphabet/adaptation/char_test.cpp:
```diff
@@ -57,19 +57,19 @@ TYPED_TEST(char_adaptation, to_char)
-    EXPECT_TRUE((std::is_same_v<decltype(assign_char_strict(TypeParam{'A'}, 'A')), alphabet_char_t<TypeParam>  >));
-    EXPECT_EQ((assign_char_strict(TypeParam{'C'}, 'A')), l);
-    EXPECT_EQ((assign_char_strict(l,              'C')), TypeParam{'C'});
+    EXPECT_TRUE((std::is_same_v<decltype(assign_char_to_strict(             'A', l)), alphabet_char_t<TypeParam> &>));
```
```suggestion
    EXPECT_TRUE((std::is_same_v<decltype(assign_char_to_strict('A', l             )), alphabet_char_t<TypeParam> &>));
```

eseiler commented on test/unit/alphabet/adaptation/char_test.cpp:
```diff
@@ -57,19 +57,19 @@ TYPED_TEST(char_adaptation, to_char)
+    EXPECT_TRUE((std::is_same_v<decltype(assign_char_to_strict(             'A', l)), alphabet_char_t<TypeParam> &>));
+    EXPECT_TRUE((std::is_same_v<decltype(assign_char_to_strict('A', TypeParam{'A'})), alphabet_char_t<TypeParam>  >));
+    EXPECT_EQ((assign_char_to_strict('A', TypeParam{'C'})), l);
+    EXPECT_EQ((assign_char_to_strict(             'C', l)), TypeParam{'C'});
```
```suggestion
    EXPECT_EQ((assign_char_to_strict('C', l             )), TypeParam{'C'});
```

eseiler replied:
Also in other places

h-2 commented:
@eseiler I have committed all your changes, fixed the warnings from the documentation and fixed some more minor whitespace issues. Can you have another look?

eseiler commented on include/seqan3/alphabet/concept.hpp:
```diff
@@ -74,13 +74,13 @@ SEQAN3_CONCEPT Semialphabet = std::Regular<std::remove_reference_t<t>> &&
-    requires           noexcept(assign_rank(std::remove_reference_t<t>{}, 0));
-    requires std::Same<decltype(assign_rank(v,                            0)), std::remove_reference_t<t> &>;
-    requires std::Same<decltype(assign_rank(std::remove_reference_t<t>{}, 0)), std::remove_reference_t<t>  >;
+    requires           noexcept(assign_rank_to(0                           , v));
```
```suggestion
    requires           noexcept(assign_rank_to(0, v                           ));
```

eseiler commented on include/seqan3/alphabet/concept.hpp:
```diff
@@ -74,13 +74,13 @@ SEQAN3_CONCEPT Semialphabet = std::Regular<std::remove_reference_t<t>> &&
-    requires std::Same<decltype(assign_rank(std::remove_reference_t<t>{}, 0)), std::remove_reference_t<t>  >;
+    requires           noexcept(assign_rank_to(0                           , v));
+    requires           noexcept(assign_rank_to(0, std::remove_reference_t<t>{}));
+    requires std::Same<decltype(assign_rank_to(0                           , v)), std::remove_reference_t<t> &>;
```
```suggestion
    requires std::Same<decltype(assign_rank_to(0, v                           )), std::remove_reference_t<t> &>;
```

eseiler commented on include/seqan3/alphabet/concept.hpp:
```diff
@@ -119,20 +119,20 @@ SEQAN3_CONCEPT Alphabet = Semialphabet<t> && requires (t v)
-    requires           noexcept(assign_char(std::remove_reference_t<t>{}, 0));
-    requires std::Same<decltype(assign_char(v,                            0)), std::remove_reference_t<t> &>;
-    requires std::Same<decltype(assign_char(std::remove_reference_t<t>{}, 0)), std::remove_reference_t<t>  >;
+    requires           noexcept(assign_char_to(0,                            v));
```
```suggestion
    requires           noexcept(assign_char_to(0, v                           ));
```

eseiler commented on include/seqan3/alphabet/concept.hpp:
```diff
@@ -119,20 +119,20 @@ SEQAN3_CONCEPT Alphabet = Semialphabet<t> && requires (t v)
-    requires std::Same<decltype(assign_char(std::remove_reference_t<t>{}, 0)), std::remove_reference_t<t>  >;
+    requires           noexcept(assign_char_to(0,                            v));
+    requires           noexcept(assign_char_to(0, std::remove_reference_t<t>{}));
+    requires std::Same<decltype(assign_char_to(0,                            v)), std::remove_reference_t<t> &>;
```
```suggestion
    requires std::Same<decltype(assign_char_to(0, v                           )), std::remove_reference_t<t> &>;
```

eseiler commented on include/seqan3/alphabet/concept.hpp:
```diff
@@ -119,20 +119,20 @@ SEQAN3_CONCEPT Alphabet = Semialphabet<t> && requires (t v)
     // strict assignment from char
-    requires std::Same<decltype(assign_char_strict(v,                            0)), std::remove_reference_t<t> &>;
-    requires std::Same<decltype(assign_char_strict(std::remove_reference_t<t>{}, 0)), std::remove_reference_t<t>  >;
+    requires std::Same<decltype(assign_char_to_strict(0,                            v)), std::remove_reference_t<t> &>;
```
```suggestion
    requires std::Same<decltype(assign_char_to_strict(0, v                           )), std::remove_reference_t<t> &>;
```

eseiler commented on include/seqan3/alphabet/detail/member_exposure.hpp:
```diff
@@ -103,14 +103,14 @@ constexpr alphabet_type & assign_rank(alphabet_type & alph, underlying_rank_t<al
  * Use this e.g. to newly create alphabet letters from rank:
  * ~~~{.cpp}
- * auto l = assign_rank(dna5{}, 1);  // l is of type dna5 and == 'C'_dna5
+ * auto l = assign_rank_to(1, dna5{});  // l is of type dna5 and == 'C'_dna5
```
Can you do a `\snippet test/snippet/alphabet/adaptation/uint.cpp assign_rank` here instead?

h-2 replied:
this will all go away / be completely changed with the next PRs anyway

eseiler commented on include/seqan3/range/view/rank_to.hpp:
```diff
@@ -51,7 +51,7 @@ namespace seqan3::view
  * | seqan3::const_iterable_concept  |                                       | *preserved*                                        |
  * |                                 |                                       |                                                    |
- * | seqan3::reference_t             | seqan3::underlying_rank_t<alphabet_t> | `alphabet_t`                                       |
+* | seqan3::reference_t             | seqan3::alphabet_rank_t<alphabet_t>   | `alphabet_t`                                       |
```
```suggestion
 * | seqan3::reference_t             | seqan3::alphabet_rank_t<alphabet_t>   | `alphabet_t`                                       |
```

eseiler commented on test/unit/alignment/configuration/align_config_edit_test.cpp:
```diff
@@ -27,9 +27,10 @@ TEST(align_cfg_edit, is_hamming)
             else
-                EXPECT_EQ((scheme.score(assign_rank(dna15{}, i), assign_rank(dna15{}, j))), -1);
+                EXPECT_EQ((scheme.score(assign_rank_to(i, dna15{}), assign_rank_to(j, dna15{}))), -1);
+
```
A new newline? 👀

h-2 replied:
will fix it

eseiler commented on test/unit/alphabet/adaptation/uint_test.cpp:
```diff
@@ -86,19 +86,19 @@ TYPED_TEST(uint_adaptation, to_char)
-    EXPECT_TRUE((std::is_same_v<decltype(assign_char_strict(TypeParam{'A'}, 'A')), underlying_rank_t<TypeParam>  >));
-    EXPECT_EQ((assign_char_strict(TypeParam{67}, 'A')), l);
-    EXPECT_EQ((assign_char_strict(l,             'C')), TypeParam{67});
+    EXPECT_TRUE((std::is_same_v<decltype(assign_char_to_strict(             'A', l)), alphabet_rank_t<TypeParam> &>));
```
```suggestion
    EXPECT_TRUE((std::is_same_v<decltype(assign_char_to_strict('A', l             )), alphabet_rank_t<TypeParam> &>));
```

eseiler commented on test/unit/alphabet/adaptation/uint_test.cpp:
```diff
@@ -86,19 +86,19 @@ TYPED_TEST(uint_adaptation, to_char)
+    EXPECT_TRUE((std::is_same_v<decltype(assign_char_to_strict(             'A', l)), alphabet_rank_t<TypeParam> &>));
+    EXPECT_TRUE((std::is_same_v<decltype(assign_char_to_strict('A', TypeParam{'A'})), alphabet_rank_t<TypeParam>  >));
+    EXPECT_EQ((assign_char_to_strict('A', TypeParam{67})), l);
+    EXPECT_EQ((assign_char_to_strict(            'C', l)), TypeParam{67});
```
```suggestion
    EXPECT_EQ((assign_char_to_strict('C', l            )), TypeParam{67});
```

h-2 commented:
@eseiler did I resolve all your issues? @smehringer can you also give it a quick review?
